amlip_py/amlip_py/node/FiwareNode.py

```diff
@@ -0,0 +1,558 @@
+            }
+        }
+
+        response = requests.post(
+            f'http://{self.context_broker_ip}:{self.context_broker_port}/v2/subscriptions',
+            headers=headers_POST,
+            data=json.dumps(subscription_data))
+
+        response.raise_for_status()
+
+    def run(self):
+        """
+        Start the Flask server.
+
+        """
+
+        self.app.run(host=self.server_ip, port=self.server_port)
+
+    def post_data(
+            self,
+            data):
+        """
+        Post data to the Fiware context broker.
+
+        Parameters
+        ----------
+        data : dict
+            Data to be posted.
+
+        """
+        global waiter_inference
+
+        # __FLAG__
+        # We want to close the waiter when we post data
+        waiter_inference = BooleanWaitHandler(True, False)
+        # __FLAG__
+
+        entity_data = {
+            'id': self.id,
+            'type': 'string',
+            self.entity_data: {
+                'value': data,
+                'type': 'InferenceDataType'
+            },
+            self.entity_solution: {
+                'value': '',
+                'type': 'InferenceSolutionDataType'
+            }
+        }
+
+        try:
+            response = requests.post(
+                f'http://{self.context_broker_ip}:{self.context_broker_port}/v2/entities',
+                headers=headers_POST,
+                data=json.dumps(entity_data))
+
+            response.raise_for_status()
+            self.logger.info('Data posted successfully')
+
+        except requests.exceptions.RequestException as e:
+            self.logger.warning(f'Failed to post data: {e} ')
+            raise e
+
+    def inference_received(
+            self,
+            inference,
+            task_id,
+            server_id):
+        """
+        Handle inference data received from the inference node.
+
+        Parameters
+        ----------
+        inference : InferenceSolutionDataType
+            Inference of the data previously sent.
+        task_id : str
+            Id of the data which this inference answers.
+        server_id : AmlipIdDataType
+            Id of the InferenceNode which sent the data.
+
+        """
+
+        self.logger.info(f'Inference received from server: {server_id}\n'
+                         f' with id: {task_id}\n'
+                         f' inference: {inference.to_string()}')
+
+        metadata = {
+            'task_id': {
+                'value': task_id,
+                'type': 'string'
+            },
+            'server_id': {
+                'value': server_id.to_string(),
+                'type': 'string'
+            }
+        }
+
+        self.patch_inference(inference.to_string(), metadata=metadata)
+        # __FLAG_
+        # We want to open it again after uploding the inference to the context broker
+        global waiter_inference
+        waiter_inference.open()
+        # __FLAG__
```

Comment:
The `waiter_inference` is being opened outside of the `patch_inference()` function. To improve reliability, it would be better to open the `waiter_inference` inside `patch_inference()` only after a successful inference patch. This way, we avoid opening it when an error occurs.

Reply:
Done